File: day3/day3.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	data, err := ioutil.ReadFile("input3.txt")
	check(err)
	visited := make(map[string]bool)
	count := 1
	x := 0
	y := 0
	visited["0-0"] = true
	key := ""
	for _, c := range data {
		cha := string(c)
		if cha == "^" {
			y += 1
		} else if cha == ">" {
			x += 1
		} else if cha == "v" {
			y -= 1
		} else if cha == "<" {
			x -= 1
		} else {
			break
		}
		key = strconv.Itoa(x) + "-" + strconv.Itoa(y)
		if visited[key] {
			// do nothing
		} else {
			count += 1
			visited[key] = true
		}
	}
	fmt.Println(count)
}

File: day3/day3.py
def switch(c, x, y, visited, count):
    if c == '<':
        x[0] -= 1
    elif c == '>':
        x[0] += 1
    elif c == '^':
        y[0] += 1
    elif c == 'v':
        y[0] -= 1
    else:
        return
    key = str(x[0]) + '-' + str(y[0])
    if key not in visited:
        visited[key] = True
        count[0] += 1


with open('input3.txt', 'r') as f:
    words = f.read()
    x, y, rx, ry = [0], [0], [0], [0]
    visited = {}
    visited[str(x[0]) + '-' + str(y[0])] = True
    count = [1]
    for i, c in enumerate(words):
        if i % 2 == 0:
            switch(c, rx, ry, visited, count)
        else:
            switch(c, x, y, visited, count)
    print count[0]
